Terms of Service for ISes and IMs (#2140)
matrix-org/matrix-doc: proposals/2140-terms-of-service-2.md

# MSC2140: Terms of Service API for Identity Servers and Integration Managers

> MSC1692 introduces a method for homeservers to require that users read and

Reviewer: ```suggestion
[MSC1692](https://github.com/matrix-org/matrix-doc/issues/1692) introduces a method for homeservers to require that users read and
```

   should not have to agree to them again when using a different client.

## Proposal

Throuhgout this proposal, $prefix will be used to refer to the prefix of the
API in question, ie. `/_matrix/identity/api/v1` for the IS API and
`/_matrix/integrations/v1` for the IM API.

This proposal introduces:
 * The `$prefix/terms` endpoint
 * The `m.third_party_terms` section in account data
 * The `X-TERMS-TOKEN` HTTP header

### Terms API

New API endpoints will be introduced:

#### `GET $prefix/terms`:
This returns a set of documents that the user must agree to abide by in order
to use the service. Its response is similar to the structure used in the
`m.terms` UI auth flow of the Client/Server API:

```json
{
    "policies": {
        "terms_of_service": {
            "version": "2.0",
            "en": {
                "name": "Terms of Service",
                "url": "https://example.org/somewhere/terms-2.0-en.html"
            },
            "fr": {
                "name": "Conditions d'utilisation",
                "url": "https://example.org/somewhere/terms-2.0-fr.html"
            }
        }
    }
}
```

> Each document (ie. key/value pair in the 'policies' object) MUST be

Reviewer 2: see thread: https://github.com/matrix-org/matrix-doc/pull/2140#discussion_r295948000

uniquely identified by its URL. It is therefore strongly recommended
that the URL contains the version number of the document. The name
and version keys, however, are used only to provide a human-readable
description of the document to the user.

In the IM API, the client should provide authentication for this endpoint.

#### `POST $prefix/terms`:
Requests to this endpoint have a single key, `user_accepts` whose value is
a list of URLs (given by the `url` field in the GET response) of documents that 
the user has agreed to:

```json
{

>     "user_accepts": ["https://example.org/somewhere/terms-2.0-en.html"]

Author: So this comes from the desire to de-dupe policies between different services, in this case by URL. If we do it by policy ID per service, we can't de-dupe which means that an org that hosts their own copy of everything still has to give the user (at least) 3 separate documents: have added this as a requirement.

This is a good point though, especially with different URLs for different languages. Hmm...

>     "user_accepts": ["https://example.org/somewhere/terms-2.0-en.html"]

Reviewer 2: ah, that does indeed make it complicated. I assume this is the case of (for example) someone wanting to use `https://example.org/terms.html` for their HS, IS, and IM? That does indeed suck if the user has to click 3 checkboxes...

Without having thought about it too much, could we add some opaque identifier to this MSC and MSC1692 to dedupe policies? Doesn't necessarily have to be a content hash, but it does allow the service admin to say that their policies are all the same, they just want the client to tell the user that. Something like:
```json
{
  "user_accepts": "B94D27B9934D3E08A52E52D7DA7DABFAC484EFE37A5380EE9088F7ACE2EFCDE9"
}
```
and
```json
{
    "policies": {
        "terms_of_service": {
            "version": "2.0",
            "id": "B94D27B9934D3E08A52E52D7DA7DABFAC484EFE37A5380EE9088F7ACE2EFCDE9"
            "en": {
                "name": "Terms of Service",
                "url": "https://example.org/somewhere/terms-2.0-en.html"
            },
            "fr": {
                "name": "Conditions d'utilisation",
                "url": "https://example.org/somewhere/terms-2.0-fr.html"
            }
        }
    }
}
```

In theory we could replace the `terms_of_service` key, but then we run into problems of potentially wanting explicit checkboxes from the user for a given thing. For example:
* [ ] I assert I am over the age of 13 as per [the terms of service]
* [ ] I have read and agree to the [terms of service]

... but then clients might dedupe the IS, IM, and HS checkboxes? This is probably less of a concern though - per-service explicit checkboxes and cross-service deduping covers most use cases, and dumb clients would just offer several checkboxes to users.

>     "user_accepts": ["https://example.org/somewhere/terms-2.0-en.html"]

Author: My only concern with this i that URLs are nicely scoped to domains, whereas with opaque IDs, *maybe* someone could copy/figure out the ID for your ToS document and use the same ID for theirs, meaning a user would never see your ToS document. Unsure how much of a concern it is, but it feels like using URLs would sidestep the issue.

>     "user_accepts": ["https://example.org/somewhere/terms-2.0-en.html"]

Reviewer 2: we could namespace the IDs? `org.example.tos.v1`?

I'm really hesitant to use URLs while we support i18n, and i18n imo is a requirement for this feature.

>     "user_accepts": ["https://example.org/somewhere/terms-2.0-en.html"]

Author: It's a bit of a faff for clients to check though (ie. parse hostname from url, split on '.', reverse array, join on '.', check ID starts with result), and security depends on clients doing that check.

Doing it by URL doesn't seem *that* bad: it does mean that the client has to do an n^2 checking each url in terms response against its list of agreed docs, but as long as it uses a sensible data structure it's not actually n^2, and we don't expect either list to be terribly large.

Another option could be to have a URL that identifies the doc (eg. `https://example.com/terms/v1` and then suffixes for each language (eg. `{"en": "/en/" }`) which would obviously restrict the URL structure - maybe that's not so bad?

Another thing in the back of my mind is how important it is that we know what language of a doc a user has agreed to in case there are discrepancies between the translations, although the c/s version of the API assumes it's not important.

>     "user_accepts": ["https://example.org/somewhere/terms-2.0-en.html"]

Reviewer 2: Forcing a URL structure sounds like a bad idea. One of the undocumented features of MSC1692 is supporting complex organizations (ie: the ones where an office in another city controls the corporate site and therefore the legal document structure). 

I'm somewhat adamant that we don't need to know the language, just that they accepted a specific document. Clients wouldn't have to parse the namespace because they'd treat it as an opaque identifier (considering the version in their de-dupe). The namespace would just be to prevent someone else also having `abcdef123` as their document ID.

>     "user_accepts": ["https://example.org/somewhere/terms-2.0-en.html"]

Author: The namespacing wouldn't work unless you enforced it somewhere though? ie. unless the client checked, there's nothing stopping me from publishing https://evil.com/empty_tos with id `org.goodguys.tos`, so if you use my IM and the good guys' IS/HS, you'll never see the ToS from the good guys (or you'll see one or the other, depending on client iteration order).

>     "user_accepts": ["https://example.org/somewhere/terms-2.0-en.html"]

Reviewer 2: hmm... maybe we don't de-dupe and just say "sorry users, click 3 checkboxes at worst"?

>     "user_accepts": ["https://example.org/somewhere/terms-2.0-en.html"]

Author: Well, if we do it by URL and the client doesn't bother to check the different-language alternatives (or indeed doesn't bother to check account data at all), this is the failure mode, (ie. the user would see a different language version of a doc they've already agreed to, or see the same doc three times).

>     "user_accepts": ["https://example.org/somewhere/terms-2.0-en.html"]

Reviewer 2: I would expect that clients only show the language which is relevant to the user anyways, regardless of what we say in the spec.

Although maybe I'm on the wrong track: the problem I'm trying to solve is inter-service duplication of policies to save the user from clicking a checkbox for the homeserver, another for the identity server, and yet another for the integration manager. If this isn't the goal, then we should be using the key name and `version` (ie: `("terms_of_service", "2.0")`) as the key for what document a user has accepted. 

The trust is in the service to not change the contents out from under the user while also giving the service an option for changing their URL structure. Services which change the contents without bumping the version number would be easily sued in court anyways.

>     "user_accepts": ["https://example.org/somewhere/terms-2.0-en.html"]

Author: Yep, that's the goal, and yeah, it's assumed that URL contents isn't going to change.

Basically, I think our options are:
 1. De-dupe by URL. Clients need to n^2 search through all the different languages to see if there's a version the user's already agreed to. If they don't, failure mode is that the user might see a document more often than they need to (not the end of the world). If the server changes URLs, the user has to re-agree.
 2. De-dupe by rDNS / other namespace identifier. Client just linear search to find an already-agreed doc. Clients must check the namespace of the identifier matches where they are fetching the doc from. If they don't, the failure mode is that a malicious HS/IS/IM could prevent a terms doc from being shown from another HS/IS/IM (potentially bad). Servers can change URLs as they wish.
3. Don't do any de-duping between services. Users always see 3 docs, even if the same org provides all 3 services (probably not an option as I think this is a requirement).

>     "user_accepts": ["https://example.org/somewhere/terms-2.0-en.html"]

Reviewer 2: I'd like Option 3 but it is a requirement to de-dupe documents for the same service. Which leaves Option 1 (current MSC) as the best option due to the security risk of Option 2. It just sucks that the language gets embedded into the acceptance.

I guess we should mirror the changes to MSC1692 for consistency?

>     "user_accepts": ["https://example.org/somewhere/terms-2.0-en.html"]

Author: What would we need to add other than that clients should add the URL of HS terms docs that they accept to `m.third_party_terms` too (maybe that it not the right name for it...) and check that before showing the user new docs?

>     "user_accepts": ["https://example.org/somewhere/terms-2.0-en.html"]

Reviewer 2: Yea, looks like 1692 doesn't need any changes. This proposal should probably clarify that the HS URLs be included in the account data field (`m.accepted_terms` instead of `m.third_party_terms`?).

There's a potential weird edge case where a client adds the HS policies to `m.third_party_terms` but the HS hasn't received the acceptance and therefore doesn't count it as accepted. Clients would have to handle this by considering the terms not accepted and prompt the user.

>     "user_accepts": ["https://example.org/somewhere/terms-2.0-en.html"]

Author: Yep, good point - added.

I don't think there's a race - `m.accepted_terms` is just the list that the client knows it's got acceptance from the user, so if the HS doesn't get the acceptance, it will error again and the client will go & check `m.accepted_terms`, see they're all accepted already and so just submit the UI auth without prompting the user.

}
```

In the IM API, the client should provide authentication for this endpoint.

> The clients MUST include the correct URL for the language of the document that

Author: If anything I'd say sending the URL that was accepted is easier than either defining a way for the client to decide which URL it sends or requiring that the server cope with the client sending any of the URLs (ie. not necessarily the one the user actually saw).

In any case it's up to the server if it wants to store it or not: if it would rather not, it can squelch all the language-variant URLs down to a single identifier internally and its privacy policy can reflect that.

> The clients MUST include the correct URL for the language of the document that

Author: Oh, and more importantly: legally I think it's important to know exactly what document a user agreed to. Unless there's a clause in the docs that states a particular language version of the doc is binding in the event of a discrepancy, the user agreed to the one they agreed to.

> The clients MUST include the correct URL for the language of the document that

Reviewer 2: For the legal side: I don't think we'll cover the thousand use cases for all the countries, but we can at least make an effort. Prior domain knowledge on my part says linking to a version (not necessarily a URL) and document name is suitable, provided the system's document management system can be audited to prove the user accepted it.

Because it seems relevant: in a prior life I spent a fair amount of time building out a terms of service + privacy policy management thing (albeit for a centralized system). The API shape was fairly simple compared to the time invested in it, but it was roughly:
* The server tracks the versions of documents (and their contents). Everything is considered a "policy document". There was also a flag for draft status (for internal editing, given the policy contents were stored local to the system).
* During registration, the server sends the current published version of the documents and asks the user to accept them. The frontend hit an API endpoint to get the policy contents, including the version. The frontend also submitted acceptance by including the version for auditing purposes. 
* During/after login was a bit more complicated. The system supported blocking and non-blocking updates to the policies (for example, typos in a document would result in a non-blocking update). The frontend handled this by showing a dialog in both cases, but if non-blocking gave the user the option of being reminded later (next startup of the app).

This is also where a lot of the inspiration for MSC1692 came from.

> The clients MUST include the correct URL for the language of the document that

Reviewer 2: The rest of my concern here is covered here: https://github.com/matrix-org/matrix-doc/pull/2140#discussion_r295948000

Both making the `POST $prefix/terms` request and providing an `X-TERMS-TOKEN`
header signal that the user consents to the terms contained within the
corresponding documents. That is to say, if a client or user obtains an
acceptance token via means other than a response to the `POST $perfix/terms`
API, inclusion of the acceptance token in an `X-TERMS-TOKEN` header in a
request still constitutes agreement to the terms in the corresponding
documents.

## Tradeoffs

This introduces a different way of accepting terms from the client/server API
which uses User-Interactive Authentication. In the client/server API, the use
of UI auth allows terms acceptance to be integrated into the registration flow
in a simple and backwards-compatible way. Indtroducing the UI Auth mechanism
into these other APIs would add significant complexity, so this functionality
has been provided with simpler, dedicated endpoints.

The `m.third_party_terms` section contains only URLs of the documents that
have been agreed to. This loses information like the name and version of
the document, but:
 * It would be up to the clients to copy this information correctly into
   account data.

>  * Having just the URLs makes it much easier for clients to make a list

Reviewer 2: it also makes it harder, as they have to search all languages in all documents to determine if a document is signed. AS hinted in previous comments, pivoting towards uniqueness on the version number+policy identifier may be safer.

Reviewer 2: see https://github.com/matrix-org/matrix-doc/pull/2140#discussion_r295948000 for acknowledgement of this limitation (and discussion around its validity)

   of URLs and find documents not already agreed to.

## Potential issues

If the server does not authentcate users, some mechanism is required to track
users agreement to terms. The introduction of an extra HTTP header on all
requests adds overhead to every request and complexity to the client to add a
custom header.

## Security considerations

The `acceptance_token` is, in effect, a cookie and could be used to identify
users of the service.  Users of the Integration manager must be authenticated
anyway, so this is irrelevant for the IM API. It could allow an Identity Server
to identify users where it may otherwise not be able to do so (if a client was
careful to mask other identifying HTTP headers). Given most requests to the IS
API, by their nature, include 3pids which, even if hashed, will make users
easily identifiable, this probably does not add any significant concern.

It is assumed that once servers publish a given version of a document at a
given URL, the contents of that URL will not change. This could be mitigated by

> identifying documents based on a hash of their contents rather than their URLs.

Reviewer 2: this gets messy when someone goes out and redesigns matrix.org though.

`/_matrix/identity/api/v1` may return a `M_TERMS_NOT_SIGNED` errcode. This
indicates that the user must agree to (new) terms in order to use or continue
to use the service.

The client uses the `GET $prefix/terms` endpoint to get the latest set of terms
that must be agreed to. It then cross-references this set of documents against
the `m.third_party_terms` account data and presents to the user any documents
that they have not already agreed to, along with UI for them to indicate their
agreement. Once the user has indicated their agreement, then, and only then,
must the client use the `POST $prefix/terms` API to signal to the server the
set of documents that the user has agreed to.

If the server returns an `acceptance_token`, the client should include this
token in the `X-TERMS-TOKEN` HTTP header in all subsequent requests to an
endpoint on the API with the exception of `/_matrix/identity/api/v1`.

Both making the `POST $prefix/terms` request and providing an `X-TERMS-TOKEN`
header signal that the user consents to the terms contained within the
corresponding documents. That is to say, if a client or user obtains an

> acceptance token via means other than a response to the `POST $perfix/terms`

Reviewer: perfix

method for homeservers to require that users read and agree to certain
documents before being permitted to use the service. This proposal introduces a
corresponding method that can be used with Identity Servers and Integration
Managers.

Requirements for this proposal are:
 * ISs and IMs should be able to give multiple documents a user must agree to
   abide by
 * Each document shoud be versioned
 * ISes and IMs must, for each request that they handle, know that the user
   making the request has agreed to their data being used. This need not be
   absolute proof (we will always have to trust that the client actually
   showed the document to the user) but it must be reasonably demonstrable that
   the user has given informed consent for the client to use that service.
 * ISs and IMs must be able to prevent users from using the service if they
   have not provided agreement.
 * A user should only have to agree to each version of each document once for
   their Matrix ID, ie. having agreed to a set of terms in one client, they
   should not have to agree to them again when using a different client.
 * Documents should be de-duplicated between services. If two or more services

>    are hosted by the same organisation, the organistation should have the

Reviewer 2: ```suggestion
   are hosted by the same organisation, the organisation should have the
```

service and so are unable to track what users have agreed to what terms in the
way that Homeservers do.

## Proposal

Throuhgout this proposal, $prefix will be used to refer to the prefix of the
API in question, ie. `/_matrix/identity/v2` for the IS API and
`/_matrix/integrations/v1` for the IM API.

Note the removal of the `/api` prefix and migration to v2 in the IS API
following convention from
[MSC2134](https://github.com/matrix-org/matrix-doc/issues/2134).

This proposal introduces:
 * The `$prefix/terms` endpoint
 * The `m.accepted_terms` section in account data

This proposal relies on both Integration Managers and Identity Servers being
able to identity users by their mxid and store the fact that a given mxid has
indicated that they accept the terms given. Integration Managers already
identity users in this way by authenticating them using the OpenID endpoint on
the Homeserver. This proposal introduces the same mechanism to Identity Servers
and adds authentication to accross the Identity Service API.

### IS API Authentication

All current endpoints within `/_matrix/identity/api/v1/` will be duplicated

> into `/_matrix/identity/v2`.

Reviewer 2: Just to clarify that MSC2134 and this MSC affect each other in a non-blocking way. The IS release wouldn't be cut without both MSCs being merged at this rate anyhow.
```suggestion
into `/_matrix/identity/v2`, noting that MSC2134 changes the behaviour of lookups. Authentication is still expected on MSC2134's proposed endpoints.
```

Any request to any endpoint within `/_matrix/identity/v2`, with the exception
of `/_matrix/identity/v2` and the new `/_matrix/identity/v2/account/register`
and `GET /_matrix/identity/v2/terms` may return an error with `M_UNAUTHORIZED`
errcode with HTTP status code 401. This indicates that the user must
authenticate with OpenID and supply a valid `access_token`.

These endpoints require authentication by the client supplying an access token
either via an `Authorization` header with a `Bearer` token or an `access_token`
query parameter.

The existing endpoints under `/_matrix/identity/api/v1/` continue to be unauthenticated.

> ISes may support the old v1 API for as long as they wish. Clients must update to use

Reviewer 2: We should explicitly say that v1 endpoints are deprecated, and how ISes may deprecate usage of the endpoints (403 M_FORBIDDEN as per my suggestion on MSC2134, imo)

From a spec perspective, we'd probably cut the next release of the IS spec with all the v1 endpoints flagged as deprecated. The release after that we'd remove them entirely.

Author: I don't think we should be using 403 for this (403 == "you can't do this but maybe you'd be able to if you were someone else") - I made this comment somewhere else recently but I'm now failing to find where.

Reviewer 2: 403 doesn't necessarily mean that someone else has access, just that permission is denied. 404 is more wrong I think because the endpoints do exist.

Reviewer 2: 401 was a better choice, indeed.

Data in the client/server API that clients SHOULD use to track what sets of
terms the user has consented to. This has an array of URLs under the 'accepted'
key to which the user has agreed to.

An `m.accepted_terms` section therefore resembles the following:

```json
{
    "accepted": [
        "https://example.org/somewhere/terms-1.2-en.html",
        "https://example.org/somewhere/privacy-1.2-en.html"
    ]
}
```

Whenever a client submits a `POST $prefix/terms` request to an IS or IM or
completes an `m.terms` flow on the HS, it SHOULD update this account data
section adding any the URLs of any additional documents that the user agreed to
to this list.

### Terms Acceptance in the API

Before any requests are made to an Identity Server or Integration Manager,
the client must use the `GET $prefix/terms` endpoint to fetch the set of
documents that the user must agree to in order to use the service.

It then cross-references this set of documents against the `m.accepted_terms`
account data and presents to the user any documents that they have not already
agreed to, along with UI for them to indicate their agreement. Once the user

> has indicated their agreement, it adds these URLs to `m.accepted_terms` account

Reviewer 2: should probably have a sentence saying that an empty `/terms` response, or no new URLs, means that the client doesn't need to do anything because there's nothing for the user to accept.

data. Once this has succeeded, then, and only then, must the client perform
OpenID authentication, getting a token from the Homeserver and submitting this
to the service using the `register` endpoint.

Having done this, if the user agreed to any new documents, it performs a `POST
$prefix/terms` request to signal to the server the set of documents that the
user has agreed to.

Any request to any endpoint in the IM API, and the `_matrix/identity/v2/`
namespace of the IS API, with the exception of `/_matrix/identity/v2` itself,
may return:

 * `M_UNAUTHORIZED` errcode with HTTP status code 401. This indicates that
   the user must authenticate with OpenID and supply a valid `access_token`.

>  * `M_CONSENT_NOT_GIVEN` errcode. This indicates that the user must agree to

Reviewer 2: Can we use M_TERMS_NOT_SIGNED to be consistent with MSC1692? https://github.com/matrix-org/matrix-doc/pull/1692/files#diff-186a5d506fe022d1dd53971801835e9dR188

Author: Oh, that explains where I got that form. So we appear to have had an implementation failure somewhere in that `M_TERMS_NOT_SIGNED` exists nowhere in either synapse or react-sdk and we've apparently used `M_CONSENT_NOT_GIVEN` instead. I'm not sure how we managed this.

Reviewer 2: M_CONSENT_NOT_GIVEN is for a different thing (https://github.com/matrix-org/matrix-doc/issues/1252), which MSC1692 replaces.

Author: Ah, I see.

   (new) terms in order to use or continue to use the service.

The `_matrix/identity/v2/3pid/unbind` must not return either of these

> errors if the request has a valid signature from a Homeserver.

Reviewer 2: Homeserver auth is optional, so the mere presence of the signature may not be enough for the IS to consider it valid (it can still refuse to do HS auth and want a direct call):
```suggestion
errors if the request has a valid signature from a Homeserver, and is being authenticated as such.
```

in the current login sessions is:

 * `GET $prefix/terms`
 * Compare result with `m.accepted_terms` account data, get set of documents
   pending agreement
 * If non-empty, show this set of documents to the user and wait for the user
   to indicate their agreement.
 * Add the newly agreed documents to `m.accepted_terms`
 * On success, or if there were no documents pending agreement, get an OpenID
   token from the Homeserver and submit this token to the `register` endpoint.
   Store the resulting access token.
 * If the set of documents pending agreement was non-empty, Perform a
   `POST $prefix/terms` request to the servcie with these documents.

## Tradeoffs

The Identity Service API previously did not require authentication, and OpenID
is reasonably complex, adding a significant burden to both clients and servers.
A custom HTTP Header was also considered that could be added to assert that the
client agrees to a particular set of terms. We decided against this in favour
of re-using existing primitives that already exist in the Matrix ecosystem.
Custom HTTP Headers are not used anywhere else within Matrix. This also gives a

> very simple and natural way for ISes to enforce that users may only bind 3pids

Reviewer 2: general nit:
* `s/3pid/3PID`
* `s/mxid/MXID`

to their own mxids.

This introduces a different way of accepting terms from the client/server API
which uses User-Interactive Authentication. In the client/server API, the use
of UI auth allows terms acceptance to be integrated into the registration flow
in a simple and backwards-compatible way. Indtroducing the UI Auth mechanism

> into these other APIs would add significant complexity, so this functionality

Reviewer 2: I'm not entirely sure this is true. We could mirror parts of UIA (like `/login` from the CS API) and still accomplish what we need. This can be a place where we say we don't have extensibility: you either return flows for terms, or no flows.

This also makes the awkward after-registration call to `POST /terms` needed only when the documents change. As an implementor, I'd probably want to deny registration in the first place if someone hasn't accepted the terms rather than let them register and hope they accept the terms later.

Author: mm, yeah - is this better?

has been provided with simpler, dedicated endpoints.

The `m.accepted_terms` section contains only URLs of the documents that
have been agreed to. This loses information like the name and version of
the document, but:
 * It would be up to the clients to copy this information correctly into
   account data.
 * Having just the URLs makes it much easier for clients to make a list
   of URLs and find documents not already agreed to.

> ## Potential issues

Reviewer 2: Many of the IS endpoints are proxied through the homeserver, which this proposal doesn't seem to address. Are homeservers expected to maintain a list of IS access tokens for their users and auth with that? (which also assumes that the HS's terms say it can do this).

Author: Right yes - I think we forgot this. I've added to put it in the same place as the id server address.

> This change is not backwards compatible: clients implementing older versions of

Reviewer 2: It is backwards compatible now: we don't touch the old endpoints beyond the standard 403 error that all clients should expect at any time anyways.

service and so are unable to track what users have agreed to what terms in the
way that Homeservers do.

## Proposal

Throuhgout this proposal, $prefix will be used to refer to the prefix of the
API in question, ie. `/_matrix/identity/v2` for the IS API and
`/_matrix/integrations/v1` for the IM API.

Note the removal of the `/api` prefix and migration to v2 in the IS API
following convention from
[MSC2134](https://github.com/matrix-org/matrix-doc/issues/2134).

This proposal introduces:
 * The `$prefix/terms` endpoint
 * The `m.accepted_terms` section in account data

This proposal relies on both Integration Managers and Identity Servers being
able to identity users by their MXID and store the fact that a given MXID has
indicated that they accept the terms given. Integration Managers already
identity users in this way by authenticating them using the OpenID endpoint on
the Homeserver. This proposal introduces the same mechanism to Identity Servers
and adds authentication to accross the Identity Service API.

### IS API Authentication

All current endpoints within `/_matrix/identity/api/v1/` will be duplicated

> into `/_matrix/identity/v2`, noting that MSC2134 changes the behaviour of lookups. Authentication is still expected on MSC2134's proposed endpoints.

Reviewer 2: wrap

Any request to any endpoint within `/_matrix/identity/v2`, with the exception

> of `/_matrix/identity/v2`, any `requestToken` endpoint and the new

Reviewer 2: this sentence has many `and`s and is begging to be a list

`/_matrix/identity/v2/account/register` and `GET /_matrix/identity/v2/terms`
may return an error with `M_UNAUTHORIZED` errcode with HTTP status code 401.
This indicates that the user must authenticate with OpenID and supply a valid
`access_token`.

`requestToken` endpoints are excluded from the auth check because they are used
in the registration process before the user has an MXID and therefore cannot
log in with OpenID. It is up to the IS to manage its privacy obligations
appropriately when fulfilling these requests, bearing in mind that the user has

> not explictly incicated their agreement to any documents, and may abort the

Reviewer 2: ```suggestion
not explicitly indicated their agreement to any documents, and may abort the
```

Identity Servers do not currently require any kind of user login to access the
service and so are unable to track what users have agreed to what terms in the
way that Homeservers do.

## Proposal

Throuhgout this proposal, $prefix will be used to refer to the prefix of the
API in question, ie. `/_matrix/identity/v2` for the IS API and
`/_matrix/integrations/v1` for the IM API.

Note the removal of the `/api` prefix and migration to v2 in the IS API
following convention from
[MSC2134](https://github.com/matrix-org/matrix-doc/issues/2134).

This proposal introduces:
 * The `$prefix/terms` endpoint
 * The `m.accepted_terms` section in account data

This proposal relies on both Integration Managers and Identity Servers being
able to identity users by their MXID and store the fact that a given MXID has
indicated that they accept the terms given. Integration Managers already
identity users in this way by authenticating them using the OpenID endpoint on
the Homeserver. This proposal introduces the same mechanism to Identity Servers
and adds authentication to accross the Identity Service API.

### IS API Authentication

> All current endpoints within `/_matrix/identity/api/v1/` will be duplicated

Author: oh, good point

into `/_matrix/identity/v2`, noting that MSC2134 changes the behaviour of
lookups. Authentication is still expected on MSC2134's proposed endpoints.

Any request to any endpoint within `/_matrix/identity/v2`, with the exception
of:
 * `/_matrix/identity/v2`
 * any `requestToken` endpoint

>  * The new `/_matrix/identity/v2/account/register` 

Reviewer 2: also logout should be here, otherwise people can't effectively delete themselves.

 * The new `GET /_matrix/identity/v2/terms`

...may return an error with `M_UNAUTHORIZED` errcode with HTTP status code 401.
This indicates that the user must authenticate with OpenID and supply a valid
`access_token`.

> `requestToken` endpoints are excluded from the auth check because they are used

Author: ah, yep

either via an `Authorization` header with a `Bearer` token or an `access_token`
query parameter.

The existing endpoints under `/_matrix/identity/api/v1/` continue to be
unauthenticated but will be deprecated.  ISes may support the old v1 API for as
long as they wish. Once ISes remove support for the old APIs, those endpoints
must return HTTP Status 404. Clients must update to use the v2 API as soon as
possible.

OpenID authentication in the IS API will work the same as in the Integration Manager
API, as specified in [MSC1961](https://github.com/matrix-org/matrix-doc/issues/1961).

When clients supply an identity server to the Homeserver in order for the Homeserver
to make calls to the IS on its behalf, it must also supply its access token for the
Identity Server alongside in the `is_token` key of the same JSON object. That is,
in the main request object for a `requestToken` request and in the `threepidCreds`
object when supplying 3PID credentials (eg. in the `m.email.identity` UI auth stage).
Exxceptions to this are any requests where the only IS operation the Homeserver may
perform is unbinding, ie. `/_matrix/client/r0/account/deactivate` and

> `/_matrix/client/r0/account/3pid/delete`, in which case the unbind is authenticated

Author: In both of these cases it's always authed by HS signature though

> `/_matrix/client/r0/account/3pid/delete`, in which case the unbind is authenticated

Reviewer 2: That's true, and the sentence does explain that. It just feels a bit too absolute, leading to the mental model of the MSC to say this endpoint isn't authenticated with OpenID. Some words to reinforce that homeserver signatures are additive rather than exclusive would help fix this.

> `/_matrix/client/r0/account/3pid/delete`, in which case the unbind is authenticated

Author: I've gone with 'will be' which hopefully captures both without adding too many extra words.

acceptance into the `acceptance_token` is not defined, eg. they may internally
transform all URLs to the URL of the English-language version of each document
if the server deems it appropriate to do so. Servers should accept agreement of
any one language of each document as sufficient, regardless of what language a
client is operating in: users should not have to re-consent to documents if
they change their client to a different language.

The server responds with an empty JSON object. The server must not assume that
the client will agree to all documents in a single request.

### Accepted Terms Account Data

This proposal also defines the `m.accepted_terms` section in User Account
Data in the client/server API that clients SHOULD use to track what sets of
terms the user has consented to. This has an array of URLs under the 'accepted'
key to which the user has agreed to.

An `m.accepted_terms` section therefore resembles the following:

```json
{
    "accepted": [
        "https://example.org/somewhere/terms-1.2-en.html",
        "https://example.org/somewhere/privacy-1.2-en.html"
    ]
}
```

Whenever a client submits a `POST $prefix/terms` request to an IS or IM or

> completes an `m.terms` flow on the HS, it SHOULD update this account data

Author: It could get confusing if sometimes the HS updates it and sometimes the client updates it so I've said do it after registration (not that having two terms account data structures isn't already confusing enough...)

section adding any the URLs of any additional documents that the user agreed to
to this list.

### Terms Acceptance in the API

Before any requests are made to an Identity Server or Integration Manager,
the client must use the `GET $prefix/terms` endpoint to fetch the set of
documents that the user must agree to in order to use the service.

It then cross-references this set of documents against the `m.accepted_terms`
account data and presents to the user any documents that they have not already
agreed to, along with UI for them to indicate their agreement. If there are no
such documents (ie. if the `policies` dict is empty or the user has already
agreed to all documents) the client proceeds to perform the OpenID
registration. Once the user has indicated their agreement, it adds these URLs
to `m.accepted_terms` account data. Once this has succeeded, then, and only

> then, must the client perform OpenID authentication, getting a token from the

Author: yeah, it's because we clarified what needs to happen if there are no new documents. I've tried to re-word it to sound less redundant.

Homeserver and submitting this to the service using the `register` endpoint.

Having done this, if the user agreed to any new documents, it performs a `POST
$prefix/terms` request to signal to the server the set of documents that the
user has agreed to.

> Any request to any endpoint in the IM API, and the `_matrix/identity/v2/`

Reviewer 2: we should also just change this to "Any endpoint with the exception of those previously mentioned may return:"

otherwise we duplicate the list, and this second list is wrong (`/register` can return M_UNAUTHORIZED to require a token? The point of the endpoint is to get that token ._. )

namespace of the IS API, with the exception of `/_matrix/identity/v2` itself,
may return:

 * `M_UNAUTHORIZED` errcode with HTTP status code 401. This indicates that
   the user must authenticate with OpenID and supply a valid `access_token`.

>  * `M_TERMS_NOT_SIGNED` errcode. This indicates that the user must agree to

Reviewer 2: 403 error code?

   (new) terms in order to use or continue to use the service.

> The `_matrix/identity/v2/3pid/unbind` must not return either of these

Reviewer 2: ```suggestion
The `/_matrix/identity/v2/3pid/unbind` endpoint must not return either of these
```

errors if the request has a valid signature from a Homeserver, and is being authenticated as such.

In summary, the process for using a service that has not previously been used
in the current login sessions is:

 * `GET $prefix/terms`
 * Compare result with `m.accepted_terms` account data, get set of documents

>    pending agreement

Reviewer 2: ```suggestion
   pending agreement.
```

   (bearing in mind that the user must be able to opt-out of components of a
   service whilst still being able to use the service without that component).

Identity Servers do not currently require any kind of user login to access the
service and so are unable to track what users have agreed to what terms in the
way that Homeservers do.

## Proposal

Throuhgout this proposal, $prefix will be used to refer to the prefix of the
API in question, ie. `/_matrix/identity/v2` for the IS API and
`/_matrix/integrations/v1` for the IM API.

Note the removal of the `/api` prefix and migration to v2 in the IS API
following convention from
[MSC2134](https://github.com/matrix-org/matrix-doc/issues/2134).

This proposal introduces:
 * A v2 API prefix, with authentication, for the Identity Service
 * The `$prefix/terms` endpoint
 * The `m.accepted_terms` section in account data
 * `POST /_matrix/client/r0/account/3pid/unbind` endpoints on the client/server
   API

This proposal removes:
 * The `bind_email` and `bind_msisdn` on the Homeserver `/register` endpoint

This proposal relies on both Integration Managers and Identity Servers being

> able to identity users by their MXID and store the fact that a given MXID has

Reviewer 3: ```suggestion
able to identify users by their MXID and store the fact that a given MXID has
```

indicated that they accept the terms given. Integration Managers already
identify users in this way by authenticating them using the OpenID endpoint on
the Homeserver. This proposal introduces the same mechanism to Identity Servers
and adds authentication across the Identity Service API.

### IS API Authentication

All current endpoints within `/_matrix/identity/api/v1/` will be duplicated
into `/_matrix/identity/v2`, noting that MSC2134 changes the behaviour of
lookups. Authentication is still expected on MSC2134's proposed endpoints.
Support for `application/x-form-www-urlencoded` parameters in requests will
be dropped from all endpoints.

Any request to any endpoint within `/_matrix/identity/v2`, with the exception
of:
 * `/_matrix/identity/v2`
 * The new `$prefix/account/register` endpoint
 * The new `GET /_matrix/identity/v2/terms`
 * `$prefix/logout`

...may return an error with `M_UNAUTHORIZED` errcode with HTTP status code 401.
This indicates that the user must authenticate with OpenID and supply a valid
`access_token`.

> All other endpoints require authentication by the client supplying an access token

Reviewer 3: I'm having a hard time following which endpoints are "all other endpoints" here. This paragraph feels like it duplicates and / or conflicts with the previous "Any request to any endpoint within `/_matrix/identity/v2`", but I am probably just reading it wrong...

Author: Hopefully clearer now?

Reviewer 3: Yes, looks better now, thanks!

`/_matrix/client/r0/account/3pid/delete`, in which case the unbind will be
authenticated by a signed request from the Homeserver.

### HS Register API

The `bind_email` and `bind_msisdn` options to `/_matrix/client/r0/register` in
the client/server API will be removed. Due to the fact that
`/_matrix/identity/v2/3pid/bind` requires authentication, it will no longer be
possible for the Homeserver to bind 3PIDs as part of the registration process.

### IS Register API

The following new APIs will be introduced to support OpenID auth as per
[MSC1961](https://github.com/matrix-org/matrix-doc/issues/1961):

 * `/_matrix/identity/v2/account/register`
 * `/_matrix/identity/v2/account`
 * `/_matrix/identity/v2/account/logout`

Note again the removal of the `/api` prefix and migration to v2 following
convention from
[MSC2134](https://github.com/matrix-org/matrix-doc/issues/2134).

### Terms API

New API endpoints will be introduced:

#### `GET $prefix/terms`:
This returns a set of documents that the user must agree to abide by in order
to use the service. Its response is similar to the structure used in the
`m.terms` UI auth flow of the Client/Server API:

```json
{
    "policies": {

>         "terms_of_service": {

Reviewer 3: I think this example would be slightly stronger if it included a second policy, so it's immediately clear there can be multiple keys under `policies`. (Of course, it does say so in the text, but showing it in the example would be nice.)

   to indicate their agreement.
 * Add the newly agreed documents to `m.accepted_terms`
 * On success, or if there were no documents pending agreement, get an OpenID
   token from the Homeserver and submit this token to the `register` endpoint.
   Store the resulting access token.
 * If the set of documents pending agreement was non-empty, Perform a
   `POST $prefix/terms` request to the servcie with these documents.

### `POST /_matrix/client/r0/account/3pid/unbind`

A client uses this client/server API endpoint to request that the Homeserver
removes the given 3PID from the given Identity Server, or all Identity Servers.
Takes parameters the same parameters as
`POST /_matrix/client/r0/account/3pid/delete`, ie. `id_server`, `medium`,
`address` and the newly added `is_token`.

Returns the same as `POST /_matrix/client/r0/account/3pid/delete`.

Clients may add IS bindings for 3PIDs that already exist on the user's
Homeserver account by using the `POST /_matrix/client/r0/account/3pid`
to re-add the 3PID.

## Tradeoffs

The Identity Service API previously did not require authentication, and OpenID
is reasonably complex, adding a significant burden to both clients and servers.

> A custom HTTP Header was also considered that could be added to assert that the

Reviewer 3: ```suggestion
A custom HTTP header was also considered that could be added to assert that the
```

client agrees to a particular set of terms. We decided against this in favour
of re-using existing primitives that already exist in the Matrix ecosystem.

> Custom HTTP Headers are not used anywhere else within Matrix. This also gives a

Reviewer 3: ```suggestion
Custom HTTP headers are not used anywhere else within Matrix. This also gives a
```

   API

This proposal removes:
 * The `bind_email` and `bind_msisdn` on the Homeserver `/register` endpoint

This proposal relies on both Integration Managers and Identity Servers being
able to identify users by their MXID and store the fact that a given MXID has
indicated that they accept the terms given. Integration Managers already
identify users in this way by authenticating them using the OpenID endpoint on
the Homeserver. This proposal introduces the same mechanism to Identity Servers
and adds authentication across the Identity Service API.

### IS API Authentication

All current endpoints within `/_matrix/identity/api/v1/` will be duplicated
into `/_matrix/identity/v2`, noting that MSC2134 changes the behaviour of
lookups. Authentication is still expected on MSC2134's proposed endpoints.
Support for `application/x-form-www-urlencoded` parameters in requests will
be dropped from all endpoints.

Any request to any endpoint within `/_matrix/identity/v2`, with the exception
of:
 * `/_matrix/identity/v2`
 * `/_matrix/identity/v2/pubkey/*`
 * The new `$prefix/account/register` endpoint
 * The new `GET /_matrix/identity/v2/terms`

>  * `$prefix/logout`

Reviewer 3: What is this `logout` here? Is the new `$prefix/account/logout` as part of the IS Register API based on MSC1961, or something else?

Reviewer 2: I believe it's the new `/account/logout` but @dbkr might have forgotten a few characters.

Author: ah, yes

...may return an error with `M_UNAUTHORIZED` errcode with HTTP status code 401.
This indicates that the user must authenticate with OpenID and supply a valid
`access_token`.

Clients authenticate either via an `Authorization` header with a `Bearer` token
or an `access_token` query parameter.

The existing endpoints under `/_matrix/identity/api/v1/` continue to be
unauthenticated but will be deprecated.  ISes may support the old v1 API for as
long as they wish. Once ISes remove support for the old APIs, those endpoints
must return HTTP Status 404. Clients must update to use the v2 API as soon as
possible.

OpenID authentication in the IS API will work the same as in the Integration Manager
API, as specified in [MSC1961](https://github.com/matrix-org/matrix-doc/issues/1961).

> When clients supply an identity server to the Homeserver in order for the

Reviewer 3: It would be helpful to also specify that the homeserver should relay the `M_TERMS_NOT_SIGNED` error for proxied requests back to the client, so that it's clear how the client should resolve the situation. This will ease some of the complexity involved in https://github.com/vector-im/riot-web/issues/10525.

and version keys, however, are used only to provide a human-readable
description of the document to the user.

This endpoint does *not* require authentication.

#### `POST $prefix/terms`:
Requests to this endpoint have a single key, `user_accepts` whose value is
a list of URLs (given by the `url` field in the GET response) of documents that 
the user has agreed to:

```json
{
    "user_accepts": ["https://example.org/somewhere/terms-2.0-en.html"]
}
```

This endpoint requires authentication.

The clients MUST include the correct URL for the language of the document that
was presented to the user and they agreed to. Servers should accept agreement
of any one language of each document as sufficient, regardless of what language
a client is operating in: users should not have to re-consent to documents if
they change their client to a different language.

The server responds with an empty JSON object. The server must not assume that
the client will agree to all documents in a single request.

### Accepted Terms Account Data

This proposal also defines the `m.accepted_terms` section in User Account
Data in the client/server API that clients SHOULD use to track what sets of

> terms the user has consented to. This has an array of URLs under the 'accepted'

Reviewer 4: Could we have a dictionary `policy_id -> policy_url` instead of an array of URLs?
That will help to know which types of document the user has already signed. Matrix clients could then detect an update of one policy.

Author: It might also be useful to know the version of the document the user saw too, so could have:

```json
{
    "accepted": [
        {
            "url": "https://example.org/somewhere/terms-1.2-en.html",
            "version": "1.2",
            "id": "terms_of_service",
        },
        {
            "url": "https://example.org/somewhere/privacy-1.2-en.html",
            "version": "1.2",
            "id": "privacy_policy",
        }
    ]
}
```

...maybe?

Reviewer 3: For this MSC, the policy ID feels like it is not meant to be trusted and could change at any time, so I am wary of making any assumptions about it. Maybe the MSC should explicitly state something like: "The terms response contains these policy IDs largely to match similar behaviour from the homeserver, but don't assume anything about them, they may change at any time, etc. The URL is the thing that matters."

Reviewer 2: the homeserver version being [MSC1692](https://github.com/matrix-org/matrix-doc/pull/1692) where the policy ID actually means something. The homeserver version supports optional and non-optional changes to policies whereas this MSC does not (for valid reasons on both fronts, I believe).

In summary, the process for using a service that has not previously been used
in the current login sessions is:

 * `GET $prefix/terms`
 * Compare result with `m.accepted_terms` account data, get set of documents
   pending agreement.
 * If non-empty, show this set of documents to the user and wait for the user
   to indicate their agreement.
 * Add the newly agreed documents to `m.accepted_terms`
 * On success, or if there were no documents pending agreement, get an OpenID
   token from the Homeserver and submit this token to the `register` endpoint.
   Store the resulting access token.
 * If the set of documents pending agreement was non-empty, Perform a
   `POST $prefix/terms` request to the service with these documents.

### `POST /_matrix/client/r0/account/3pid/unbind`

A client uses this client/server API endpoint to request that the Homeserver
removes the given 3PID from the given Identity Server, or all Identity Servers.
Takes parameters the same parameters as
`POST /_matrix/client/r0/account/3pid/delete`, ie. `id_server`, `medium`,
`address` and the newly added `is_token`.

Returns the same as `POST /_matrix/client/r0/account/3pid/delete`.

> Clients may add IS bindings for 3PIDs that already exist on the user's

Reviewer 3: The default state going forward (in Riot at least) will be that users first add a 3PID to the HS only. Then, in a separate step, they can also choose to share it with the IS as well.

It would be ideal if we could bind to the IS via the HS for a 3PID already known to HS (which is what will soon be happening quite often) in a single operation. Currently though, using the `POST /_matrix/client/r0/account/3pid` flow to re-add a 3PID with binding will fail with `M_THREEPID_IN_USE`. (It actually fails during the earlier `POST /_matrix/client/r0/account/3pid/email/requestToken` step.)

Since these CS APIs are authed, the HS knows whether the 3PID is in use by the authed user. It would be best to change the spec to allow re-adding an in use 3PID if it's in use by the current user.

In the short term, Riot will work around this by using two step operation of removing the 3PID from the HS and then re-adding with bind true.

Reviewer 3: There are several different user flows involving 3PIDs, including:

1. adding a 3PID during registration
2. supplying a 3PID for password reset
3. adding a 3PID to your account later on for recovery
4. sharing a 3PID for discovery

It would best from a data flow and privacy perspective if flows 1 - 3 could be all be done using the HS only, possibly delegating to token sending server if needed. https://github.com/matrix-org/synapse/issues/5751 is heading in that direction.

For case 4 of sharing a 3PID for discovery, we currently proxy via the HS to the IS for this, mainly so that the HS has a record of 3PID associations that can be unbound when an account is deactivated.

![Untitled](https://user-images.githubusercontent.com/279572/62789655-504a1e80-bac1-11e9-91a3-f608c3d00872.png)

With v2 IS APIs and terms requirements, it becomes more complex for the HS to do this proxying. The user may need to agree terms with the IS, but the client isn't communicating with the IS directly.

@dbkr has written up [several different options](https://gist.github.com/dbkr/77ea4baf646031b4b2edfcc88998d9a9) of how we could move forward. Option 4 there:

> 4. Do the bind directly with the IS. The HS then won't know about the binding, but we could abuse the `onbind` callback to notify the HS that the 3pid has been bound. The `onbind` is unauthenticated so would have to be authed somehow (ensuring we had signoff from the *user* to add the binding: we can't just auth the IS or anyone could set up an IS and  fill your account with random bindings). The onbind callback is also not necessarily super reliable: sydent will still do the bind successfully if it can't hit the onbind callback.

seems appealing, but it's so far unclear how to resolve the authentication issue.

Reviewer 3: At the moment, we are leaning towards option 2:

> 2. Re-use `/account/3pid` (ie. the add-3pid API) with the bind flag set. This requires that HSes allow the requestToken to proceed if the 3pid is already registered to your account (synapse currently denies if the 3pid is known on the HS at all).

Reviewer 3: [MSC2229](https://github.com/matrix-org/matrix-doc/blob/anoa/rebind_3pids/proposals/2229-rebind-existing-3pid.md) now describes this.

Reviewer 3: Maybe this proposal should be updated to reference the new MSC as the suggest way of rebinding?

Reviewer 2: (in general, please link to the proposal PR instead of the markdown, as people delete branches sometimes: https://github.com/matrix-org/matrix-doc/pull/2229 )

Reviewer: I'm struggling to follow if this thread has been resolved or not?

Reviewer 3: I believe the technical problem is solved by #2229. It would be nice to mention that in this proposal, so I believe that's the only action left to resolve this thread.

## Proposal

Throuhgout this proposal, $prefix will be used to refer to the prefix of the
API in question, ie. `/_matrix/identity/v2` for the IS API and
`/_matrix/integrations/v1` for the IM API.

Note the removal of the `/api` prefix and migration to v2 in the IS API
following convention from
[MSC2134](https://github.com/matrix-org/matrix-doc/issues/2134).

This proposal introduces:
 * A v2 API prefix, with authentication, for the Identity Service
 * The `$prefix/terms` endpoint
 * The `m.accepted_terms` section in account data
 * `POST /_matrix/client/r0/account/3pid/unbind` endpoints on the client/server
   API

This proposal removes:
 * The `bind_email` and `bind_msisdn` on the Homeserver `/register` endpoint

This proposal relies on both Integration Managers and Identity Servers being
able to identify users by their MXID and store the fact that a given MXID has
indicated that they accept the terms given. Integration Managers already
identify users in this way by authenticating them using the OpenID endpoint on
the Homeserver. This proposal introduces the same mechanism to Identity Servers
and adds authentication across the Identity Service API.

### IS API Authentication

> All current endpoints within `/_matrix/identity/api/v1/` will be duplicated

Reviewer 5: Do we also want to drop lookups done without using the hashing aware endpoints?

Author: yeah, in reality we'll probably end up having non-hashed on v1 and hashed on v2, then deprecating non-hashed and v1 at the same time. I didn't wamt to tie this too much to the hashing msc so not sure whether to add that here or not.

Reviewer 6: v2 supports non-hashed so there's no reason to keep v1 around.

Any request to any endpoint within `/_matrix/identity/v2`, with the exception
of:
 * `/_matrix/identity/v2`
 * `/_matrix/identity/v2/pubkey/*`
 * The new `$prefix/account/register` endpoint
 * The new `GET /_matrix/identity/v2/terms`
 * `$prefix/account/logout`

...may return an error with `M_UNAUTHORIZED` errcode with HTTP status code 401.
This indicates that the user must authenticate with OpenID and supply a valid
`access_token`.

Clients authenticate either via an `Authorization` header with a `Bearer` token
or an `access_token` query parameter.

The existing endpoints under `/_matrix/identity/api/v1/` continue to be
unauthenticated but will be deprecated.  ISes may support the old v1 API for as
long as they wish. Once ISes remove support for the old APIs, those endpoints
must return HTTP Status 404. Clients must update to use the v2 API as soon as
possible.

OpenID authentication in the IS API will work the same as in the Integration Manager
API, as specified in [MSC1961](https://github.com/matrix-org/matrix-doc/issues/1961).

When clients supply an identity server to the Homeserver in order for the
Homeserver to make calls to the IS on its behalf, it must also supply its

> access token for the Identity Server alongside in the `is_token` key of the

Reviewer 2: `identity_access_token` doesn't seem horrible as a not-boolean, not-vague answer

> access token for the Identity Server alongside in the `is_token` key of the

Reviewer 3: Right, that seems like a fine alternative as well.

> access token for the Identity Server alongside in the `is_token` key of the

Reviewer 3: (I do wish the server field was `identity_server` instead of `id_server` for consistency, but too late for that problem.)

> access token for the Identity Server alongside in the `is_token` key of the

Author: yeah, I like this better.

> access token for the Identity Server alongside in the `is_token` key of the

Reviewer 2: nothing wrong with a different MSC which says "deprecate this in favour of a more sensible thing". The fallback is relatively straightforward, but there is bikeshed potential. Either way, a concern for a different MSC on the `id_server` front I think

same JSON object. That is, in the main request object for a `requestToken`
request and in the `threepidCreds` object when supplying 3PID credentials (eg.

> in the `m.email.identity` UI auth stage). The server must also relay

Reviewer 3: Should the `is_token` also be passed for 3PID invites via `/_matrix/client/r0/rooms/{roomId}/invite`?

Reviewer 6: I would also like to know this.

Author: Yep, it probably should. Will update.

> `M_TERMS_NOT_SIGNED` errors back to the client. Exceptions to this are any requests

Author: I've deliberately avoided putting this info in error responses to keep the terms in one place, just on the terms endpoint. The client would still have to know which provider it needed to send the POST request to in order to accept the terms.

### IS Register API

The following new APIs will be introduced to support OpenID auth as per
[MSC1961](https://github.com/matrix-org/matrix-doc/issues/1961):

 * `/_matrix/identity/v2/account/register`
 * `/_matrix/identity/v2/account`
 * `/_matrix/identity/v2/account/logout`

Note again the removal of the `/api` prefix and migration to v2 following
convention from
[MSC2134](https://github.com/matrix-org/matrix-doc/issues/2134).

### Terms API

New API endpoints will be introduced:

#### `GET $prefix/terms`:
This returns a set of documents that the user must agree to abide by in order
to use the service. Its response is similar to the structure used in the
`m.terms` UI auth flow of the Client/Server API:

```json
{
    "policies": {
        "terms_of_service": {
            "version": "2.0",
            "en": {
                "name": "Terms of Service",
                "url": "https://example.org/somewhere/terms-2.0-en.html"
            },
            "fr": {
                "name": "Conditions d'utilisation",
                "url": "https://example.org/somewhere/terms-2.0-fr.html"
            }
        },
        "privacy_policy": {
            "version": "1.2",

>             "en": {

Reviewer 5: Do we want to support having internationalisation done by the website? e.g. a single `https://matrix.org/terms-2.0.html` that would Do The Right Thing based on lang headers or geo-ip or whatever? Or in that case do we just have a single `"en"` key anyway as clients will just fall back to that?

Author: Unsure - is there precedent for this in the HS version? Either way I vote we follow that for now and change both together if we want to.

Reviewer 5: Oh, true. Yeah, lets keep it consistent.

                "name": "Privacy Policy",

>                 "url": "https://example.org/somewhere/privaacy-1.2-en.html"

Reviewer 6: ```suggestion
                "url": "https://example.org/somewhere/privacy-1.2-en.html"
```

            },
            "fr": {
                "name": "Politique de confidentialité",
                "url": "https://example.org/somewhere/privacy-1.2-fr.html"
            }
        }
    }
}
```

Each document (ie. key/value pair in the 'policies' object) MUST be

> uniquely identified by its URL. It is therefore strongly recommended

Reviewer 5: Would it make sense to uniquely identify by tuple of `(<type>, <version>, <url>)`? That would give de-duplication while still allowing just `https://matrix.org/legal.html`, as well as not having to do any crazy `n²` lookups?

Author: Possibly - otoh I sort of like forcing the version to be in the URL for general URL transparency.

Reviewer 5: Yeah, I'm just a little nervous that if we end up having three or more different services, each with multiple policies, with some churn of versions of the years, plus an increasing number of languages, then we'll end up with quite a lot of different stuff to check.

Author: Travis points out that we discussed this kind of thing in https://github.com/matrix-org/matrix-doc/pull/2140#discussion_r295948000 as eventually the conclusion was that de-duping by URL is probably the lesser evil.

agreement, then once the user has indicated their agreement, it adds these URLs
to `m.accepted_terms` account data and then proceeds with OpenID
authentication, getting a token from the Homeserver and submitting this to the
service using the `register` endpoint.

Having done this, if the user agreed to any new documents, it performs a `POST
$prefix/terms` request to signal to the server the set of documents that the
user has agreed to.

Any request to any endpoint in the IM API, and the `/_matrix/identity/v2/`
namespace of the IS API, with the exception of `/_matrix/identity/v2` itself,
may return:

 * `M_UNAUTHORIZED` errcode with HTTP status code 401. This indicates that
   the user must authenticate with OpenID and supply a valid `access_token`.
 * `M_TERMS_NOT_SIGNED` errcode with HTTP status code 403. This indicates
   that the user must agree to (new) terms in order to use or continue to
   use the service.

The `/_matrix/identity/v2/3pid/unbind` endpoint must not return either of these
errors if the request has a valid signature from a Homeserver, and is being authenticated as such.

In summary, the process for using a service that has not previously been used

> in the current login sessions is:

Reviewer 6: ```suggestion
in the current login session is:
```

 * `GET $prefix/terms`
 * Compare result with `m.accepted_terms` account data, get set of documents
   pending agreement.
 * If non-empty, show this set of documents to the user and wait for the user
   to indicate their agreement.

>  * Add the newly agreed documents to `m.accepted_terms`

Reviewer 6: ```suggestion
 * Add the newly agreed documents to `m.accepted_terms`.
```

 * If the set of documents pending agreement was non-empty, Perform a
   `POST $prefix/terms` request to the servcie with these documents.

### `POST /_matrix/client/r0/account/3pid/unbind`

A client uses this client/server API endpoint to request that the Homeserver
removes the given 3PID from the given Identity Server, or all Identity Servers.
Takes parameters the same parameters as

> `POST /_matrix/client/r0/account/3pid/delete`, ie. `id_server`, `medium`,

Reviewer 2: Should we just deprecate the delete endpoint?

Author: It's still needed for clients to remove the 3pid from the *HS* though?
